swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -465,66 +359,85 @@ extension Adapter: CallbackHandlerDelegate {
-        self.logger.error("Adapter.onUpdateResources: No network settings")
-        return
-      }
-      let updatedResourceDomains = networkResources.compactMap { $0.resourceLocation.domain }
-      networkSettings.setResourceDomains(updatedResourceDomains)
-      if case .tunnelReady = self.state {
-        networkSettings.apply(on: packetTunnelProvider, logger: self.logger, completionHandler: nil)
-      }
+      // Update resource list
+      displayableResources.update(resources: networkResources.map { $0.displayableResource })
     }
   }
 
   public func onDisconnect(error: String) {
     workQueue.async { [weak self] in
       guard let self = self else { return }
 
-      self.logger.error(
+      logger.log("\(#function)")
+
+      // Unexpected disconnect initiated by connlib. Typically for 401s.
+      logger.error(
         "Connlib disconnected with unrecoverable error: \(error)")
-      switch self.state {
-      case .stoppingTunnel(session: _, let onStopped):
-        onStopped?()
-        self.state = .stoppedTunnel
-      case .stoppingTunnelTemporarily(session: _, let onStopped):
-        onStopped?()
-        self.state = .stoppedTunnel
-      case .stoppedTunnel:
-        // This should not happen
-        break
-      case .stoppedTunnelTemporarily:
-        self.state = .stoppedTunnel
-      default:
-        packetTunnelProvider?.handleTunnelShutdown(
-          dueTo: .connlibDisconnected,
-          errorMessage: error)
-        self.packetTunnelProvider?.cancelTunnelWithError(
-          AdapterError.connlibFatalError(error))
-        self.state = .stoppedTunnel
-      }
+      packetTunnelProvider?.handleTunnelShutdown(
+        dueTo: .connlibDisconnected,
+        errorMessage: error)
+      // TODO: Define more connlib error types across the FFI so we can switch on them
+      // more granularly, and not just sign the user out every time this is called.
+      packetTunnelProvider?.cancelTunnelWithError(
+        AdapterError.connlibFatalError(error))
     }
   }
+
+  public func getSystemDefaultResolvers() -> String {
+    #if os(macOS)
+      let resolvers = SystemConfigurationResolvers(logger: logger).getDefaultDNSServers()
+    #elseif os(iOS)
+      let resolvers = resetToSystemDNSGettingBindResolvers()
+    #endif
+
+    logger.log("\(#function): \(resolvers)")
+
+    return try! String(
+      decoding: JSONEncoder().encode(resolvers),
+      as: UTF8.self
+    )
+  }
 }
+
+// MARK: Getting System Resolvers on iOS
+#if os(iOS)
+  extension Adapter {
+    // When the tunnel is up, we can only get the system's default resolvers
+    // by reading /etc/resolv.conf when matchDomains is set to a non-empty string.
+    // If matchDomains is an empty string, /etc/resolv.conf will contain connlib's
+    // sentinel, which isn't helpful to us.
+    private func resetToSystemDNSGettingBindResolvers() -> [String] {
+      var resolvers: [String] = []
+      let semaphore = DispatchSemaphore(value: 0)
```

Comment:
Each path update callback spawns a thread (`workQueue.async`), so we're ok to block here with a semaphore.

Reply:
Actually, I noticed this sometimes blocked the main UI thread on iOS, so I enclosed this inside a `Task` in the calling function.